File: 589_N_ary_Tree_PreOrder_Traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> values;
        return preorderTraversal(root, values); 
    }
    
    vector<int> preorderTraversal(Node* root, vector<int> values){
        if(root == NULL) return values;
        values.push_back(root->val);
        if(root->children.size() > 0){
            for(int i = 0; i < root->children.size(); i++){
                values = preorderTraversal(root->children[i], values); 
            }
        }
        return values;
    }
};

File: 589_N_ary_Tree_PreOrder_Traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        return self.preorderHelper(root, [])
    
    def preorderHelper(self, root, values): 
        if root == None:
            return values 
        values.append(root.val)
        if root.children != None:
            for i in root.children:
                values = self.preorderHelper(i, values)
        return values 
        